### PR TITLE
Fix issue #2446 - Enable device updates after restart

### PR DESCRIPTION
**Pull request by DigitalBites**
Lots more details at the link below - this is to fix the issue of having to restart the binding after openHab2 restarts. Will provide a pull request after I've completed my testing.

https://community.openhab.org/t/wemo-insight-switch-not-initalized-after-start-openhab2/15921/10?u=digitalbites

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@hmerk Could you review and comment please?

**Comment by DigitalBites** [comment]
The build here fails because of the test case: 
```
assert that thing handles REFRESH command correctly(org.eclipse.smarthome.binding.wemo.handler.test.WemoHandlerOSGiTest)  Time elapsed: 10.811 sec  <<< FAILURE!
java.lang.AssertionError:
Expected: is <ONLINE>
     but: was <OFFLINE>
```

This is caused by lines that i've updated in WemoHandler.java - comment out lines fixes it but i'm not sure it's 'correct':
```
private Runnable refreshRunnable = new Runnable() {

        @Override
        public void run() {
            try {
                if (!isUpnpDeviceRegistered()) {
                    logger.warn("WeMo UPnP device {} not yet registered", getUDN());
                    // updateStatus(ThingStatus.OFFLINE);
                    // return;
                }

                updateWemoState();
                onSubscription();

            } catch (Exception e) {
                logger.debug("Exception during poll : {}", e);
            }
        }
    };
```

I intentionally set the device as 'offline' until it registered via Upnp and the Upnp framework is responding. I didn't see any event updates until that happened. Not sure if the test case should be updated or the code should altered.

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 1]
```diff
@@ -81,7 +83,15 @@
...
             try {
+                if (!isUpnpDeviceRegistered()) {
+                    logger.warn("WeMo UPnP device {} not yet registered", getUDN());
+                    updateStatus(ThingStatus.OFFLINE);
+                    return;
+                }
+
                 updateWemoState();
+                onSubscription();
+
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: I don't think it is a good idea to add subscriptions on every polling run, so please remove it here

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 2]
```diff
@@ -338,44 +351,75 @@ public void onValueReceived(String variable, String value, String service) {
...
             ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+            String subscription = null;
 
             if (thingTypeUID.equals(THING_TYPE_INSIGHT)) {
-                service.addSubscription(this, "insight1", SUBSCRIPTION_DURATION);
+                subscription = "insight1";
+
             } else {
+                subscription = "basicevent1";
                 service.addSubscription(this, "basicevent1", SUBSCRIPTION_DURATION);
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: You are going to subscribe twice (L369) so please remove L364

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 3]
```diff
@@ -338,44 +351,75 @@ public void onValueReceived(String variable, String value, String service) {
...
+            String subscription = null;
 
             if (thingTypeUID.equals(THING_TYPE_INSIGHT)) {
-                service.removeSubscription(this, "insight1");
+                subscription = "insight1";
+
             } else {
-                service.removeSubscription(this, "basicevent1");
+                subscription = "basicevent1";
+                service.addSubscription(this, "basicevent1", SUBSCRIPTION_DURATION);
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: Please remove this line, as we are removing subscriptions here, not adding

**Comment by hmerk** [comment]
@DigitalBites thanks for your contribution. Please see my first comments above. I might come up with further comments.
Best
Hans-Jörg

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 4]
```diff
@@ -338,44 +351,75 @@ public void onValueReceived(String variable, String value, String service) {
...
             } else {
+                subscription = "basicevent1";
                 service.addSubscription(this, "basicevent1", SUBSCRIPTION_DURATION);
             }
+
+            if ((subscriptionState.get(subscription) == null) || !subscriptionState.get(subscription).booleanValue()) {
+                logger.debug("Setting up GENA subscription {}: Subscribing to service {}...", getUDN(), subscription);
+                service.addSubscription(this, subscription, SUBSCRIPTION_DURATION);
+                subscriptionState.put(subscription, true);
+            }
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: As you are running onSubscription on every polling run, you better check if there is no active subscription before adding a new one.

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 5]
```diff
@@ -81,7 +83,15 @@
         @Override
         public void run() {
             try {
+                if (!isUpnpDeviceRegistered()) {
+                    logger.warn("WeMo UPnP device {} not yet registered", getUDN());
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: please reduce to debug logging level

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 6]
```diff
@@ -338,44 +351,75 @@ public void onValueReceived(String variable, String value, String service) {
...
-        if (service.isRegistered(this)) {
-            if (refreshJob == null || refreshJob.isCancelled()) {
-                Configuration config = getThing().getConfiguration();
-                int refreshInterval = DEFAULT_REFRESH_INTERVAL;
-                Object refreshConfig = config.get("refresh");
-                if (refreshConfig != null) {
-                    refreshInterval = ((BigDecimal) refreshConfig).intValue();
-                }
-                refreshJob = scheduler.scheduleAtFixedRate(refreshRunnable, 0, refreshInterval, TimeUnit.SECONDS);
+        // if (service.isRegistered(this)) {
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: please remove commented line

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java [hunk 7]
```diff
@@ -338,44 +351,75 @@ public void onValueReceived(String variable, String value, String service) {
...
+            int refreshInterval = DEFAULT_REFRESH_INTERVAL;
+            Object refreshConfig = config.get("refresh");
+            if (refreshConfig != null) {
+                refreshInterval = ((BigDecimal) refreshConfig).intValue();
             }
+            refreshJob = scheduler.scheduleAtFixedRate(refreshRunnable, 0, refreshInterval, TimeUnit.SECONDS);
         }
+        // } else {
+        // logger.error("onUpdate FAILED for '{}' service.isRegistered(this) was FALSE", this);
+        // }
```
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: same for lines 416-418

Comment (DigitalBites) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: done

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java (hunk 5 above)
Comment (DigitalBites) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: done

File: extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java (hunk 4 above)
Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: I don't see any check of subscriptionState before adding a new subscription. Am I missing something?

Comment (hmerk) [inline extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java]: Thanks for clarifying, i saw this checks, but was not shure about how the boolean check is handled at all.

**Comment by hmerk** [comment]
@DigitalBites thanks for your updates. Could you add your changes to the other handlers as well and then squash your commits.

@kaikreuzer after squashing, LGTM.

**Comment by DigitalBites** [comment]
@hmerk  I believe I have squashed all of my commits although I am not a Git expert. Please review and let me know otherwise it should be set.

**Comment by hmerk** [comment]
@DigitalBites 
I still see 8 commits, so you did not squash them.
What about my question fo adopting your changes to the other handlers in the WeMo binding?

**Comment by hmerk** [comment]
@DigitalBites Any news to my last comment ?

**Comment by DigitalBites** [comment]
@hmerk Haven't forgotten. Got busy with work and the fix is working. I have to still figure out the github squashing and eventually update the other handlers. Hopefully can pick this back up in a few days.

**Comment by hmerk** [comment]
@DigitalBites thanks for your answer.
If it would be easier for you, we could close this PR, as I implemented your changes in my fork and already adopted those for the WemoLightHandler. Nothing needed to change in WemoLinkHandler and WemoMakerHandler.
So a single commit would be ready just by pressing a button.
Thanks again for your contribution.

**Comment by DigitalBites** [comment]
@hmerk still traveling on this end... I'm fine with you having this in your fork as long as it makes it back here for all of us and so i don't have to keep compiling my version. do you know when your branch will be integrated with MASTER? If not I'll clean up my loco repo and try and figure out what i did wrong with squashing my commits.

**Comment by hmerk** [comment]
As I said, once we close this PR here, I just need to press a button and we have mine redy for merging. It is then up to Kai when he could do the final review for merging. Benefit would be that I also implemented your changes for the WmoLightHandler.

**Comment by kaikreuzer** [comment]
@hmerk Sounds good - let's close this PR and you create a new one (please add Jayson with an "Also-by" line in the commit message).

**Pull request closed without merging**